### DISCUSSION TITLE
Java 17 compatibility

### DISCUSSION THREAD

**Discussion by BenoitMathiaud** [DominoKit/domino-ui#922] [General]
Hi Guys,

Thank you for the great work on this library.
I'm just starting to discover it but seems a good continuation for the GWT philosophy.

Question about JDK compatibility, I failed to compile the CRM sample application with Java 17 : _ClassNotFoundException: com.google.gwt.dev.jjs.ast.JPrimitiveType_
**But it works fine with Java 11.**
The Getting Started prerequisites says : Java 11 or higher ... ?

What are the compatible JDK levels to use Domino UI 2.0.0 ?

Thanks,
Ben

**Comment by niloc132** [comment]
Can you share the full build log, and confirm what GWT version you are using? This looks like the kind of error we see when someone has mixed their classpath with more than one copy of GWT, or different parts of GWT from different versions.

Latest GWT (and domino) should build cleanly with at least Java 11-21.

**Comment by vegegoku** [comment]
It seems I had an updated branch locally that I never pushed, you can pull and update then try, as @niloc132 said it should work with java version 11-21 no issues.

**Comment by BenoitMathiaud** [comment]
Thanks for quick replies.

I reproduce the issue with the latest version of the CRM of today.
I'm using : 
- JDK 17.0.7
- GWT 2.11 form the pom.xml

Here is output of the **mvn clean verify** : 

"C:\Program Files\Semeru\jdk-17.0.7.7-openj9\bin\java.exe" -Dmaven.multiModuleProjectDirectory=E:\WS\tec\webapps\crm-main -Djansi.passthrough=true "-Dmaven.home=C:\Program Files (x86)\JetBrains\IntelliJ IDEA Community Edition 2023.3.3\plugins\maven\lib\maven3" "-Dclassworlds.conf=C:\Program Files (x86)\JetBrains\IntelliJ IDEA Community Edition 2023.3.3\plugins\maven\lib\maven3\bin\m2.conf" "-Dmaven.ext.class.path=C:\Program Files (x86)\JetBrains\IntelliJ IDEA Community Edition 2023.3.3\plugins\maven\lib\maven-event-listener.jar" "-javaagent:C:\Program Files (x86)\JetBrains\IntelliJ IDEA Community Edition 2023.3.3\lib\idea_rt.jar=57067:C:\Program Files (x86)\JetBrains\IntelliJ IDEA Community Edition 2023.3.3\bin" -Dfile.encoding=UTF-8 -classpath "C:\Program Files (x86)\JetBrains\IntelliJ IDEA Community Edition 2023.3.3\plugins\maven\lib\maven3\boot\plexus-classworlds-2.7.0.jar;C:\Program Files (x86)\JetBrains\IntelliJ IDEA Community Edition 2023.3.3\plugins\maven\lib\maven3\boot\plexus-classworlds.license" org.codehaus.classworlds.Launcher -Didea.version=2023.3.4 clean verify
[INFO] Scanning for projects...
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Build Order:
[INFO] 
[INFO] crm                                                                [pom]
[INFO] crm-shared                                                         [jar]
[INFO] crm-client                                                     [gwt-app]
[INFO] crm-server                                                         [war]
[INFO] 
[INFO] ---------------------< org.dominokit.samples:crm >----------------------
[INFO] Building crm 1.0-SNAPSHOT                                          [1/4]
[INFO]   from pom.xml
[INFO] --------------------------------[ pom ]---------------------------------
[INFO] 
[INFO] --- clean:3.2.0:clean (default-clean) @ crm ---
[INFO] 
[INFO] ------------------< org.dominokit.samples:crm-shared >------------------
[INFO] Building crm-shared 1.0-SNAPSHOT                                   [2/4]
[INFO]   from crm-shared\pom.xml
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- clean:3.2.0:clean (default-clean) @ crm-shared ---
[INFO] 
[INFO] --- resources:3.3.1:resources (default-resources) @ crm-shared ---
[INFO] Copying 3 resources from src\main\java to target\classes
[INFO] 
[INFO] --- compiler:3.11.0:compile (default-compile) @ crm-shared ---
[INFO] Changes detected - recompiling the module! :source
[INFO] Compiling 3 source files with javac [debug target 11] to target\classes
[WARNING] system modules path not set in conjunction with -source 11
[INFO] 
[INFO] --- resources:3.3.1:testResources (default-testResources) @ crm-shared ---
[INFO] skip non existing resourceDirectory E:\WS\tec\webapps\crm-main\crm-shared\src\test\resources
[INFO] 
[INFO] --- compiler:3.11.0:testCompile (default-testCompile) @ crm-shared ---
[INFO] No sources to compile
[INFO] 
[INFO] --- surefire:3.1.2:test (default-test) @ crm-shared ---
[INFO] No tests to run.
[INFO] 
[INFO] --- jar:3.3.0:jar (default-jar) @ crm-shared ---
[INFO] Building jar: E:\WS\tec\webapps\crm-main\crm-shared\target\crm-shared-1.0-SNAPSHOT.jar
[INFO] 
[INFO] --- source:3.0.1:jar-no-fork (attach-sources) @ crm-shared ---
[INFO] Building jar: E:\WS\tec\webapps\crm-main\crm-shared\target\crm-shared-1.0-SNAPSHOT-sources.jar
[INFO] 
[INFO] ------------------< org.dominokit.samples:crm-client >------------------
[INFO] Building crm-client 1.0-SNAPSHOT                                   [3/4]
[INFO]   from crm-client\pom.xml
[INFO] ------------------------------[ gwt-app ]-------------------------------
[INFO] 
[INFO] --- clean:3.2.0:clean (default-clean) @ crm-client ---
[INFO] 
[INFO] --- gwt:1.0.1:enforce-encoding (default-enforce-encoding) @ crm-client ---
[INFO] 
[INFO] --- gwt:1.0.1:add-super-sources (default-add-super-sources) @ crm-client ---
[INFO] 
[INFO] --- gwt:1.0.1:add-test-super-sources (default-add-test-super-sources) @ crm-client ---
[INFO] 
[INFO] --- gwt:1.0.1:generate-module (default-generate-module) @ crm-client ---
[INFO] 
[INFO] --- resources:3.3.1:resources (default-resources) @ crm-client ---
[INFO] skip non existing resourceDirectory E:\WS\tec\webapps\crm-main\crm-client\src\main\resources
[INFO] skip non existing resourceDirectory E:\WS\tec\webapps\crm-main\crm-client\src\main\super
[INFO] 
[INFO] --- compiler:3.11.0:compile (default-compile) @ crm-client ---
[INFO] Changes detected - recompiling the module! :dependency
[INFO] Compiling 21 source files with javac [debug target 11] to target\classes
[WARNING] system modules path not set in conjunction with -source 11
[INFO] /E:/WS/tec/webapps/crm-main/crm-client/src/main/java/org/dominokit/samples/Shell.java: E:\WS\tec\webapps\crm-main\crm-client\src\main\java\org\dominokit\samples\Shell.java uses or overrides a deprecated API.
[INFO] /E:/WS/tec/webapps/crm-main/crm-client/src/main/java/org/dominokit/samples/Shell.java: Recompile with -Xlint:deprecation for details.
[INFO] /E:/WS/tec/webapps/crm-main/crm-client/src/main/java/org/dominokit/samples/CrmPage.java: E:\WS\tec\webapps\crm-main\crm-client\src\main\java\org\dominokit\samples\CrmPage.java uses unchecked or unsafe operations.
[INFO] /E:/WS/tec/webapps/crm-main/crm-client/src/main/java/org/dominokit/samples/CrmPage.java: Recompile with -Xlint:unchecked for details.
[INFO] 
[INFO] --- resources:3.3.1:testResources (default-testResources) @ crm-client ---
[INFO] skip non existing resourceDirectory E:\WS\tec\webapps\crm-main\crm-client\src\test\resources
[INFO] skip non existing resourceDirectory E:\WS\tec\webapps\crm-main\crm-client\src\test\super
[INFO] 
[INFO] --- compiler:3.11.0:testCompile (default-testCompile) @ crm-client ---
[INFO] No sources to compile
[INFO] 
[INFO] --- gwt:1.0.1:test (default-test) @ crm-client ---
[INFO] No tests to run.
[INFO] 
[INFO] --- gwt:1.0.1:compile (default-compile) @ crm-client ---
[INFO] Compiling module org.dominokit.samples.App
[INFO]    [ERROR] Unexpected internal compiler error
[INFO] java.lang.RuntimeException: Unexpected error deserializing AST for 'java.lang.Throwable'
[INFO] 	at com.google.gwt.dev.javac.CompilationUnit.getTypes(CompilationUnit.java:373)
[INFO] 	at com.google.gwt.dev.jjs.impl.UnifyAst.assimilateSourceUnit(UnifyAst.java:1022)
[INFO] 	at com.google.gwt.dev.jjs.impl.UnifyAst.internalFindType(UnifyAst.java:1483)
[INFO] 	at com.google.gwt.dev.jjs.impl.UnifyAst.translate(UnifyAst.java:1528)
[INFO] 	at com.google.gwt.dev.jjs.impl.UnifyAst.resolveType(UnifyAst.java:1440)
[INFO] 	at com.google.gwt.dev.jjs.impl.UnifyAst.assimilateSourceUnit(UnifyAst.java:1033)
[INFO] 	at com.google.gwt.dev.jjs.impl.UnifyAst.internalFindType(UnifyAst.java:1483)
[INFO] 	at com.google.gwt.dev.jjs.impl.UnifyAst.translate(UnifyAst.java:1528)
[INFO] 	at com.google.gwt.dev.jjs.impl.UnifyAst.resolveType(UnifyAst.java:1440)
[INFO] 	at com.google.gwt.dev.jjs.impl.UnifyAst.assimilateSourceUnit(UnifyAst.java:1033)
[INFO] 	at com.google.gwt.dev.jjs.impl.UnifyAst.internalFindType(UnifyAst.java:1483)
[INFO] 	at com.google.gwt.dev.jjs.impl.UnifyAst.translate(UnifyAst.java:1528)
[INFO] 	at com.google.gwt.dev.jjs.impl.UnifyAst.resolveType(UnifyAst.java:1440)
[INFO] 	at com.google.gwt.dev.jjs.impl.UnifyAst.assimilateSourceUnit(UnifyAst.java:1033)
[INFO] 	at com.google.gwt.dev.jjs.impl.UnifyAst.internalFindType(UnifyAst.java:1483)
[INFO] 	at com.google.gwt.dev.jjs.impl.UnifyAst.findType(UnifyAst.java:1466)
[INFO] 	at com.google.gwt.dev.jjs.JavaToJavaScriptCompiler.synthesizeEntryMethodHolderInit(JavaToJavaScriptCompiler.java:1347)
[INFO] 	at com.google.gwt.dev.jjs.JavaToJavaScriptCompiler.unifyJavaAst(JavaToJavaScriptCompiler.java:1402)
[INFO] 	at com.google.gwt.dev.jjs.JavaToJavaScriptCompiler.constructJavaAst(JavaToJavaScriptCompiler.java:1222)
[INFO] 	at com.google.gwt.dev.jjs.JavaToJavaScriptCompiler.precompile(JavaToJavaScriptCompiler.java:1140)
[INFO] 	at com.google.gwt.dev.jjs.JavaToJavaScriptCompiler.precompile(JavaToJavaScriptCompiler.java:255)
[INFO] 	at com.google.gwt.dev.Precompile.precompile(Precompile.java:243)
[INFO] 	at com.google.gwt.dev.Precompile.precompile(Precompile.java:190)
[INFO] 	at com.google.gwt.dev.Precompile.precompile(Precompile.java:131)
[INFO] 	at com.google.gwt.dev.Compiler.compile(Compiler.java:192)
[INFO] 	at com.google.gwt.dev.Compiler.compile(Compiler.java:143)
[INFO] 	at com.google.gwt.dev.Compiler.compile(Compiler.java:132)
[INFO] 	at com.google.gwt.dev.Compiler$1.run(Compiler.java:110)
[INFO] 	at com.google.gwt.dev.CompileTaskRunner.doRun(CompileTaskRunner.java:55)
[INFO] 	at com.google.gwt.dev.CompileTaskRunner.runWithAppropriateLogger(CompileTaskRunner.java:50)
[INFO] 	at com.google.gwt.dev.Compiler.main(Compiler.java:113)
[INFO] Caused by: java.lang.ClassNotFoundException: com.google.gwt.dev.jjs.ast.JPrimitiveType
[INFO] 	at java.base/java.lang.Class.forNameImpl(Native Method)
[INFO] 	at java.base/java.lang.Class.forName(Class.java:429)
[INFO] 	at java.base/java.io.ClassByNameCache$FutureValue.get(ClassByNameCache.java:188)
[INFO] 	at java.base/java.io.ClassByNameCache.get(ClassByNameCache.java:159)
[INFO] 	at java.base/java.io.ObjectInputStream.resolveClass(ObjectInputStream.java:891)
[INFO] 	at java.base/java.io.ObjectInputStream.readNonProxyDesc(ObjectInputStream.java:2153)
[INFO] 	at java.base/java.io.ObjectInputStream.readClassDesc(ObjectInputStream.java:2017)
[INFO] 	at java.base/java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2343)
[INFO] 	at java.base/java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1852)
[INFO] 	at java.base/java.io.ObjectInputStream$FieldValues.<init>(ObjectInputStream.java:2732)
[INFO] 	at java.base/java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:2581)
[INFO] 	at java.base/java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2376)
[INFO] 	at java.base/java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1852)
[INFO] 	at java.base/java.io.ObjectInputStream.readObject(ObjectInputStream.java:591)
[INFO] 	at java.base/java.io.ObjectInputStream.readObject(ObjectInputStream.java:501)
[INFO] 	at java.base/java.util.ArrayList.readObject(ArrayList.java:899)
[INFO] 	at jdk.internal.reflect.GeneratedMethodAccessor31.invoke(Unknown Source)
[INFO] 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[INFO] 	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
[INFO] 	at java.base/java.io.ObjectStreamClass.invokeReadObject(ObjectStreamClass.java:1100)
[INFO] 	at java.base/java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:2547)
[INFO] 	at java.base/java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2376)
[INFO] 	at java.base/java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1852)
[INFO] 	at java.base/java.io.ObjectInputStream.readObject(ObjectInputStream.java:591)
[INFO] 	at java.base/java.io.ObjectInputStream.readObject(ObjectInputStream.java:501)
[INFO] 	at com.google.gwt.dev.jjs.ast.JDeclaredType.readMembers(JDeclaredType.java:550)
[INFO] 	at com.google.gwt.dev.jjs.ast.JProgram.deserializeTypes(JProgram.java:237)
[INFO] 	at com.google.gwt.dev.javac.CompilationUnit.getTypes(CompilationUnit.java:369)
[INFO] 	... 30 more
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for crm 1.0-SNAPSHOT:
[INFO] 
[INFO] crm ................................................ SUCCESS [  0.110 s]
[INFO] crm-shared ......................................... SUCCESS [  1.954 s]
[INFO] crm-client ......................................... FAILURE [ 48.700 s]
[INFO] crm-server ......................................... SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  51.819 s
[INFO] Finished at: 2024-03-21T15:27:27+01:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal net.ltgt.gwt.maven:gwt-maven-plugin:1.0.1:compile (default-compile) on project crm-client: GWT exited with status 1 -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
[ERROR] 
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <args> -rf :crm-client

Process finished with exit code 1

**Comment by niloc132** [comment]
Thanks for the additional detail. This is a known problem with the IBM JDK, their ObjectOutputStream/ObjectInputStream is apparently non-standard, but you can force it to be standard with `-Dcom.ibm.enableClassCaching=false` passed as a jvm arg to the compiler. See https://github.com/gwtproject/gwt/issues/9789 for more detail.

Alternatively, use an openjdk implementation instead, which will not have this problem.

**Reply by niloc132** [reply]
Also see https://github.com/eclipse-openj9/openj9/issues/9912 - this has apparently been known for a long time on their side.

**Comment by BenoitMathiaud** [comment]
I switched to OpenJDK and it works fine.
Thank you !
